botocore/utils.py

```diff
@@ -794,5 +800,117 @@ def _switch_hosts(request, new_endpoint, use_new_scheme=True):
         ''
     )
     final_endpoint = urlunsplit(final_endpoint_components)
-    logger.debug('Updating URI from %s to %s' % (request.url, final_endpoint))
-    request.url = final_endpoint
+    logger.debug('Updating URI from %s to %s' % (
+        original_endpoint, final_endpoint))
+    return final_endpoint
+
+
+class S3RegionRedirector(object):
+    def __init__(self, endpoint_bridge, cache=None):
+        self._endpoint_resolver = endpoint_bridge
+        self._cache = cache or {}
+        self._cache_lock = threading.Lock()
+
+    def _add_bucket_to_cache(self, bucket, signing_context):
+        with self._cache_lock:
+            self._cache[bucket] = signing_context
+
+    def _get_bucket_from_cache(self, bucket):
+        with self._cache_lock:
+            signing_context = self._cache.get(bucket, None)
+        return signing_context
+
+    def register(self, emitter):
+        emitter.register('needs-retry.s3', self.redirect_from_error)
+        emitter.register('before-call.s3', self.set_request_url)
+        emitter.register('before-parameter-build.s3', self.redirect_from_cache)
+
+    def redirect_from_error(self, request_dict, response, **kwargs):
+        """
+        An S3 request sent to the wrong region will return an error that
+        contains the endpoint the request should be sent to. This handler
+        will add the redirect information to the signing context and then
+        redirect the request.
+        """
+        if response is None:
+            # This could be none if there was a ConnectionError or other
+            # transport error.
+            return
+
+        error = response[1].get('Error', {})
+        if error.get('Code') != 'PermanentRedirect':
+            return
+
+        if 'context' not in request_dict:
```

Comment:
Same thing as above.  Unless there's a strong reason to have context be optional, it simplifies code if make make guarantees about the contents of the request dict.